src/parse.cpp

```diff
@@ -74,7 +76,16 @@ parser::function_guard parser::enter_function(function_attributes attributes) {
     this->in_generator_function_ = false;
     break;
   }
-  return function_guard(this, was_in_async_function, was_in_generator_function);
+  this->in_loop_statement_ = false;
+  this->in_switch_statement_ = false;
+  return function_guard(this, was_in_async_function, was_in_generator_function,
+                        was_in_loop_statement, was_in_switch_statement);
+}
+
+parser::loop_guard parser::enter_loop() {
+  bool was_in_loop_statement = this->in_loop_statement_;
+  this->in_loop_statement_ = true;
+  return loop_guard(this, was_in_loop_statement);
```

Comment:
Nit: We can be fancy and use std::exchange:
```suggestion
  return loop_guard(this, std::exchange(this->in_loop_statement_, true));
```